### ISSUE TITLE
pyslurm leaking memory

### ISSUE THREAD

**Issue by giovtorres** [gingergeeks/pyslurm#60]
I have a [script](https://gist.github.com/giovtorres/6581bb529437eed7c64c) that runs in an infinite loop, gets the cpu and memory utilization of the cluster, sends it to a local network socket, sleeps for 30 seconds, and then repeats.

Running it this way appears to leak memory.  The RSS memory usage for the script continuously grows.  After running the script through Valgrind for a few minutes, I get the following summary which indicates a memory leak:

```
==1609== HEAP SUMMARY:
==1609==     in use at exit: 4,528,140 bytes in 155,013 blocks
==1609==   total heap usage: 23,995,476 allocs, 23,840,463 frees, 6,139,163,824 bytes alloc
==1609== LEAK SUMMARY:
==1609==    definitely lost: 3,134,362 bytes in 142,471 blocks
==1609==    indirectly lost: 0 bytes in 0 blocks
==1609==      possibly lost: 47,002 bytes in 759 blocks
==1609==    still reachable: 1,346,776 bytes in 11,783 blocks
==1609==         suppressed: 0 bytes in 0 blocks
==1609== 
==1609== ERROR SUMMARY: 114 errors from 114 contexts (suppressed: 61 from 9)
```

In the while loop, I re-run the get_cluster_util() function, which in turn calls pyslurm.node().get() and pyslurm.job().get() during every loop.  The get() method runs __load() and __get(), which, as far as I know, is the only way to update the node/job information for respective dictionaries that get returned.

Do you have any suggestions on how to structure scripts using pyslurm in 'while True' loops that don't leak memory?

Thanks,
Giovanni

**Comment by gingergeeks** [comment]
It means I've not cleanly up memory correctly somewhere and so we need to locate and fix that ! Thank you for the report, when I get time I will investigate as this is important to fix.

**Comment by giovtorres** [comment]
Command:
valgrind -v --leak-check=full --show-reachable=yes /usr/bin/python send_cluster_util.py

Output:
[valgrind - output - slurm_cluster_util.txt](https://github.com/gingergeeks/pyslurm/files/75670/valgrind.-.output.-.slurm_cluster_util.txt)

**Comment by gingergeeks** [comment]
Giovanni,
               Thank you for this report and your patience ! I will switch from your other request to try and track this down.  I don't use valgrind that much but clearly I should.

Mark

**Comment by giovtorres** [comment]
No worries.  Happy to help!

**Comment by gingergeeks** [comment]
Please do, if we can nail down the area of the leak. Is it a string copy, not free'ing an old pointer etc.

**Comment by gingergeeks** [comment]
Found one possible area which has made things better (not all reports fixed yet) in class job to do with slurm hostlists that I was allocating/using but calling destroy method when finished.

**Comment by gingergeeks** [comment]
Corrected another possible leak in the class job and valgrind is now not reporting any definite/indirect lost bytes in a job query loop with lots of jobs coming and going.

    ==16469== LEAK SUMMARY:
    ==16469==    definitely lost: 0 bytes in 0 blocks
    ==16469==    indirectly lost: 0 bytes in 0 blocks
    ==16469==      possibly lost: 31,405 bytes in 571 blocks

  The RSS does increase very very slowly and not steadily and may be Python increases memory usage due to increases in dictionary sizes etc.

**Comment by giovtorres** [comment]
That looks much better!  I haven't had a chance yet to look at the code, but I did see a lot of hostlist references in the stack traces from valgrind.

**Comment by giovtorres** [comment]
[valgrind docs](http://valgrind.org/docs/manual/quick-start.html#quick-start.interpret) have a helpful summary about interpreting the output:

There are several kinds of leaks; the two most important categories are:
- **definitely lost**: your program is leaking memory -- fix it!
- **probably lost**: your program is leaking memory, unless you're doing funny things with pointers (such as moving them to point to the middle of a heap block).

**Comment by gingergeeks** [comment]
That's useful, I was trying the copious amount of python interpreter output and following [https://github.com/cython/cython/wiki/UsingValgrindToDebug Cython-Using Valgrind]

It may be that because we are using the Python interpreter that we are seeing "probably lost" values. Just firing up python in valgrind and shutting it down will report some "probably lost" values.

**Comment by giovtorres** [comment]
Let me know when you have a patch ready.  I can test it. 

I was originally using this inside collectd, as it is compiled with a python interpreter.  Collectd, at least for me, normally uses under 100MB of RSS.  When running this script under collectd, RSS ballooned to 1GB within an hour and OOM-killed within a day :(

I then change the script to push the metrics directly to graphite, bypassing collectd altogether.  The memory leak was not as pronounced, but slowly and linearly increases.

So, I can test and find out quickly.

**Comment by gingergeeks** [comment]
I also need to check the node object for potential issues, is tomorrow okay ?

**Comment by gingergeeks** [comment]
Okay,,,no promises here as this is just the job class being modified slightly (pyslurm-14.11.5). I expect that we may be leaking in the node class and in other places until I take a look at the code tomorrow.
[pyslurm.pyx.zip](https://github.com/gingergeeks/pyslurm/files/75770/pyslurm.pyx.zip)

**Comment by gingergeeks** [comment]
A quick change to the node class as well in case you are still leaking (very probably).

Best regards

Mark
[pyslurm.pyx.zip](https://github.com/gingergeeks/pyslurm/files/75777/pyslurm.pyx.zip)

**Comment by giovtorres** [comment]
Thanks. Ill give it a try.
We can pick this up next week.

**Comment by gingergeeks** [comment]
Update - a refactor of various code paths as I found other potential leaks. Please try this version as I ran it overnight with a modified version of your script and it appeared to not be leaking. 

[pyslurm.pyx.zip](https://github.com/gingergeeks/pyslurm/files/76260/pyslurm.pyx.zip)

**Comment by gingergeeks** [comment]
Giovanni, Did that update resolve any long running leaks ?

**Comment by giovtorres** [comment]
I started the same script this morning.  So far it looks good.  I'm letting it run overnight...

**Comment by giovtorres** [comment]
The memory is still growing, but very very slowly.  I ran my script through valgrind yesterday and didn't get and definitely/indirectly lost bytes.   I also didn't see the same pyslurm messages in the output as before.  It might be something else I am doing. I'll keep watching.  I'll close this ticket for now.

Thanks for your help.

**Comment by gingergeeks** [comment]
Ok Giovanni but we need to track it down, Is it usable or still not usable for long running processes ?

**Comment by giovtorres** [comment]
I'm not fully convinced. Let me look at it some more....

**Comment by gingergeeks** [comment]
I'm still looking into this stripping out everywhere except the basic load method

**Comment by gingergeeks** [comment]
If I change the underlying load method in the node class just to do the initialise and then free -

      last_time = <time_t>NULL
      errCode = slurm.slurm_load_node(last_time, &new_node_info_ptr, 0)
      slurm.slurm_free_node_info_msg(new_node_info_ptr)

Fire up Python via valgrind and perform the minimal load method when the node class object is created -

     >>> import pyslurm
     >>> a = pyslurm.node()

Then I still show possibly lost blocks via valgrind and at this point I've not done much at all....still investigating.

**Comment by giovtorres** [comment]
I rebuilt a version of python 2.7.5 without pymalloc and with valgrind support.  I'm re-running my tests... i'll let it run overnight to see how much the memory grows. No matter which version of python I use, I seem to always get possibly lost blocks, even if I have a very basic script. 

After freeing the hostlist patch [hl.destroy()], the definitely lost and indirectly lost blocks have been 0 consistently.  So, i'm not sure how to interpret the current output from valgrind.

**Comment by giovtorres** [comment]
These have since been addressed in b239161, 54b67eec and e9b1c3d.
